### PR TITLE
A0-4294: New network building code

### PR DESCRIPTION
**Pull request by timorleph**
# Description

New network building code to replace substrate's `build_network`. Not yet used, there is a separate task for that. It might have some minor problems due to being untested, but I'll fix them when actually coding the usage. Also discovered and fixed some other small problems in previous unused code while implementing this.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

# Checklist:

- I have made corresponding changes to the existing documentation
- I have created new documentation

### CODE REVIEW

File: finality-aleph/src/network/build/transactions.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+) -> Result<TransactionsHandlerController<TP::Hash>, NetworkError> {
+    let (transaction_service, transaction_interface) = transactions_prototype.build(
+        network,
+        sync_service,
+        Arc::new(TransactionPoolAdapter::new(transaction_pool, client)),
+        metrics_registry,
+    )?;
+    spawn_handle.spawn(
+        "network-transactions-handler",
+        Some("networking"),
```
Comment (kostekIV) [inline finality-aleph/src/network/build/transactions.rs:53]: this - `Some("networking")` is present in a lot of spawns - extract it to const.

File: finality-aleph/src/network/build/mod.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    pub authentication_network: ProtocolNetwork,
+    pub block_sync_network: ProtocolNetwork,
+    // names chosen for compatibility with SpawnTaskParams, get better ones if we ever stop using that
+    pub sync_service: Arc<SyncingService<TP::Block>>,
+    pub tx_handler_controller: TransactionsHandlerController<TP::Hash>,
+    pub system_rpc_tx: TracingUnboundedSender<RpcRequest<TP::Block>>,
+}
+
+/// Start everything necessary to run the inter-node network and return the interfaces for it.
+// TODO(A0-3576): This code should be used.
```
Comment (kostekIV) [inline finality-aleph/src/network/build/mod.rs:48]: Pls add to the comment what are the things startet up here

**Review by kostekIV — APPROVED** [review]

File: finality-aleph/src/network/base_protocol/service.rs [hunk 3]
```diff
@@ -70,7 +70,7 @@ where
         protocol_names: Vec<ProtocolName>,
         network: Arc<NetworkService<B, BlockHash>>,
         events_from_network: Box<dyn NotificationService>,
-    ) -> (Self, SyncingService<B>) {
+    ) -> (Self, Arc<SyncingService<B>>) {
```
[the remark this answers is not in the document]
Comment (timorleph) [reply finality-aleph/src/network/base_protocol/service.rs:73]: Yup, that's a bit annoying, but all the usage places require the `Arc` so it felt better to return it here.

More annoyingly the `Arc` is completely pointless – `SyncingService` is inherently thread safe and cloneable, so all the substrate functions could have just taken it. :/

File: finality-aleph/src/network/build/transactions.rs (hunk 1 above)
Comment (timorleph) [reply finality-aleph/src/network/build/transactions.rs]: Good idea, done.

File: finality-aleph/src/network/build/mod.rs (hunk 2 above)
Comment (timorleph) [reply finality-aleph/src/network/build/mod.rs]: Done.

**Review by woocash2 — APPROVED** [review]
Lookin good

**Review by fixxxedpoint — APPROVED** [review]

**Pull request merged**